pkg/api/core/v1/client/apps.go

```diff
@@ -413,3 +415,79 @@ func (c *Client) AppExec(namespace string, appName string, tty kubectlterm.TTY)
 
 	return tty.Safe(fn)
 }
+
+type PortForwardOpts struct {
+	Address      []string
+	Ports        []string
+	StopChannel  chan struct{}
+	ReadyChannel chan struct{}
+	Out          io.Writer
+	ErrOut       io.Writer
+}
+
+func NewPortForwardOpts(address, ports []string) *PortForwardOpts {
+	opts := &PortForwardOpts{
+		Address:      address,
+		Ports:        ports,
+		StopChannel:  make(chan struct{}),
+		ReadyChannel: make(chan struct{}),
+		Out:          os.Stdin,
+		ErrOut:       os.Stderr,
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
+
+	go func() {
+		<-signals
+		if opts.StopChannel != nil {
+			close(opts.StopChannel)
+		}
+	}()
+
+	return opts
+}
+
+// AppPortForward will forward the local traffic to a remote app
+func (c *Client) AppPortForward(namespace string, appName string, opts *PortForwardOpts) error {
+	endpoint := fmt.Sprintf("%s%s/%s", c.URL, api.WsRoot, api.WsRoutes.Path("AppPortForward", namespace, appName))
+	portForwardURL, err := url.Parse(endpoint)
+	if err != nil {
+		return err
+	}
+
+	if err := c.addAuthTokenToUrl(portForwardURL); err != nil {
+		return err
+	}
+
+	upgradeRoundTripper := NewUpgrader(spdy.RoundTripperConfig{
+		TLS:                      http.DefaultTransport.(*http.Transport).TLSClientConfig, // See `ExtendLocalTrust`
+		FollowRedirects:          true,
+		RequireSameHostRedirects: false,
+		PingPeriod:               time.Second * 5,
+	})
+
+	wrapper := transport.NewBasicAuthRoundTripper(c.user, c.password, upgradeRoundTripper)
+
+	dialer := gospdy.NewDialer(upgradeRoundTripper, &http.Client{Transport: wrapper}, "GET", portForwardURL)
+	fw, err := portforward.NewOnAddresses(dialer, opts.Address, opts.Ports, opts.StopChannel, opts.ReadyChannel, opts.Out, opts.ErrOut)
+	if err != nil {
+		return err
+	}
+
+	return fw.ForwardPorts()
+}
+
+func (c *Client) addAuthTokenToUrl(url *url.URL) error {
```

Comment:
Note: 
Hm, this doesn't just add the auth token to the URL, it also makes a whole new GET request to fetch the auth token. So the func name does not suggest that.

I probably would not do this extraction for the sake of de-duplication.  I preferred the direct return of values (instead of modifying a pointer arg) and having the request logic/error handling in the top level function.

On the other hand it seems wrong to repeat the `values.Add("authtoken", token)` logic. That's an implementation detail of the auth endpoint 🤷

Reply:
The drawback is that if we are going to change the auth logic (i.e. using a JWT, adding an header, or even just changing the query param name) then we would need to change it everywhere.

The main reason of not returning the value is that in this case you are kind of telling the caller that you have probably created a new struct and you should rely on that one. And also you can keep it shorter checking the error in one line.

For ref the oauth library for example is modifying the request in order to add the token:
https://github.com/golang/oauth2/blob/d3ed0bb246c8d3c75b63937d9a5eecff9c74d7fe/token.go#L74-L81

Reply:
I was thinking that we could "split" the logic, adding a function to the Client like `Authenticate` that will do the request and saving the token (or wich auth method we will decide to use) and then a `SetAuthenticationToken` to the request, so at least the fetch/set logic is separate and the flow will be something like

```
func (c *Client) addAuthTokenToUrl(url *url.URL) error {
    token, err := c.AuthToken()
    if err != nil {
	return err
    }

    c.authToken = token
    return nil
}

func (c *Client) SetAuth(url *url.URL) {
    values := url.Query()
    values.Add("authtoken", token)
    url.RawQuery = values.Encode()
}
```

```
if err := c.Authenticate(); err != nil {
	return err
}

c.SetAuth(portForwardURL)
```

Reply:
> The drawback is that if we are going to change the auth logic (i.e. using a JWT, adding an header, or even just changing the query param name) then we would need to change it everywhere.

Well, yes, but "everywhere" is only two places right now. The auth token can only be used for websocket connections. 
The client code does a lot of this low level stuff, e.g. like the instance query param is handled or setting headers. 

But I'm fine with the addAuthTokenToURL ([URL](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)) extraction, it's a bit surprising and the name is a bit long, but the content is short enough and it's not widely used.

As a method on the client I wouldn't want to signal that this can be used for auth in general. Maybe we need a websocket client if we use more of them.